### ISSUE TITLE
set_property changes the protoset instead of the properties

### ISSUE THREAD

**Issue by snipercup** [peter-kish/gloot#150]
I have a pistol magazine that starts with 19 bullets. When I shoot a bullet, all magazines will have 18 bullets remaining, not just the one I used to fire.

I can see here that properties returns a dictionary of size 0:
![image](https://github.com/peter-kish/gloot/assets/50166150/d5cf2036-061e-404b-9945-a38c1333c7e8)

Here you can see that the protoset is modified (the default current_ammo value is 19):
![image](https://github.com/peter-kish/gloot/assets/50166150/14881951-a0a1-46fe-bf64-91ab3471e532)

This is where I use set_property:
![image](https://github.com/peter-kish/gloot/assets/50166150/9dc8abd8-dcc0-44e3-b8ae-2785021c9765)

Here I am using both get and set property. I decided to add current_ammo to the item definition, so line 177 to 181 are never called now
![image](https://github.com/peter-kish/gloot/assets/50166150/48aaf8c0-18f3-4637-b5f0-bbd55c547559)

I am also getting this in the output tab, but I don't know if it's related. I think this started when I started using get_property and set_property
![image](https://github.com/peter-kish/gloot/assets/50166150/584f2898-f0c2-4b7d-b557-4a56af3801da)

As I understand it, using get_property and set_property should not modify the protoset, only the properties of the InventoryItem. For context, I am using a custom item editor that stores data in json. When the game starts, all item json is written to item_protosets.tres, which is then used in-game. I would like the option to set my item json file instead of the protoset, but I'll submit a different issue for that.

What is wrong here? I can't figure it out.

**Comment by snipercup** [comment]
If I access the properties data directly, the same thing happens
``` GDScript
func find_compatible_magazine(oldMagazine: InventoryItem) -> InventoryItem:
	var inventoryItems: Array = myInventory.get_items()  # Retrieve all items in the inventory

	#print_debug("find_compatible_magazine. inventory items = " + str(inventoryItems.size()))
	for item in inventoryItems:
		# Check if the item is a magazine and is compatible with the equipped weapon
		print_debug("find_compatible_magazine  not item.get_property(Magazine) == null = " + str( not item.get_property("Magazine") == null))
		if item.properties.has("Magazine") and not item == oldMagazine:
			var magazine = item.properties["Magazine"]
			#if oldMagazine:
				#print_debug("item = " + str(item.name) + ", oldMagazine = " + str(oldMagazine.name))
			if magazine.has("current_ammo"):
				#print_debug("find_compatible_magazine not item == myMagazine = " + str(not item == oldMagazine))
				if int(magazine.current_ammo) > 0:
					print_debug("magazine.current_ammo = " + str(magazine.current_ammo))
					return item
		else: 
			var itemData = Gamedata.get_data_by_id(Gamedata.data.items, "pistol_magazine")
			var magazinedata = itemData.Magazine
			item.properties["Magazine"] = magazinedata
			#if magazine.has("max_ammo"):
				#magazine.current_ammo = magazine.max_ammo
				#
				#print_debug("setting property magazine on object = " + str(item.name))
				#item.set_property("Magazine", magazine)
	return null  # Return null if no compatible magazine is found
```

**Comment by snipercup** [comment]
I think I found a workaround by adding .duplicate() after the get_property function
```GDScript
func _subtract_ammo(amount: int):
	var magazine: InventoryItem = equipmentSlot.get_magazine()
	if magazine:
		var magazineProperties = magazine.get_property("Magazine").duplicate()
		var ammunition: int = int(magazineProperties["current_ammo"])
		ammunition -= amount
		magazineProperties["current_ammo"] = ammunition
		print_debug("setting property magazine on object = " + str(magazine.name))
		magazine.set_property("Magazine", magazineProperties)
		ammo_changed.emit(get_current_ammo(), get_max_ammo(), equipped_left)
```

This is what was happening:
1. I get the property from the magazine
2. InventoryItem cannot find the "Magazine" property in properties, so it will get the "Magazine" property from the protoset
3. The "Magazine" property is returned, but this is an array, so instead of the value, it returns a reference to the array in the protoset
4. I was manipulating the magazine array, and thus the protoset.

I don't think it should be this easy to edit the protoset. My suggestion is to change the `get_item_property` function in `item_protoset.gd` to return a duplicate if it is an array or an object

**Comment by peter-kish** [comment]
So yeah, Godot arrays, dictionaries and `Object` inheritants are passed by reference and everything else is passed by value. 

That's one of the reasons why I tried to disallow those types in protosets, but I guess there are still ways to add them by e.g. defining a protoset as a JSON (I don't think I added any checks for that when parsing a JSON). I'm suspecting this is the case because the "Magazine" property of your item is a dictionary, but the protoset editor does not support dictionaries:
![Screenshot_20240201_214654](https://github.com/peter-kish/gloot/assets/11254996/30786f9c-3b97-42c2-8334-4ba297964223)

Another thing that seems a bit weird is that you have a magazine item that has a "Magazine" property, which is a dictionary that holds magazine information. I don't know what your protoset looks like, but from my perspective it would make more sense to define one prototype for magazines (e.g. with the prototype ID "pistol_magazine") which holds multiple individual properties (e.g. "current_ammo", "max_ammo" etc.). Decreasing the current ammo would as simple as:
```gdscript
if magazine.get_property("current_ammo") > 0:
    magazine.set_property("current_ammo", magazine.get_property("current_ammo") - 1)
```
and since there are no dictionaries/arrays, there's no need to worry about references.

P.S. I'm still thinking about supporting dictionaries and arrays in protosets, but because of the increased complexity I don't think it will happen before version 3.0.0.

**Comment by snipercup** [comment]
Thank you for the response. I did not know there was no support for dictionaries and arrays. My item definitions have standard weight and volume and name and descriptions. Their type is defined by the existence of a certain key, in this case "Magazine". I could add current_ammo as a flat property, but then how would I tell the difference between a magazine and a gun, which could also have current_ammo? I could have a property called "type", and give it the "Magazine" type, but what if I also want it to be a tool? My item definitions are going to be complex and will probably need nested dictionaries in the future. 

You are right about not worrying about dictionaries and arrays if I only use flat properties, but that will not support my use case. The workaround is working for now. I can see why the arrays and dictionaries will add complexities. I am not really asking to fully support them, but it would be nice if a duplicate is returned when the property is pulled from the protoset.

**Comment by peter-kish** [comment]
> how would I tell the difference between a magazine and a gun, which could also have current_ammo?

That's what prototype IDs are meant for. E.g.:
```json
[
    {
        "id": "magazine",
        "current_ammo": 16,
        "max_ammo": 20
    },
    {
        "id": "gun",
        "current_ammo": 10
    }
]
```

Once unique IDs are defined, you can use them to get items from the inventory with `get_item_by_id` or `get_items_by_id`. E.g.
```gdscript
var magazine := inventory.get_item_by_id("magazine")
var gun := inventory.get_item_by_id("gun")
if magazine && gun:
    # reload gun using magazine or something...
```

> but what if I also want it to be a tool?

Oooh, ok. So, any item would be able to have a set of these "traits"... Interesting design, but I'm somewhat unsure how well the current state of protosets fits into it - it's ID based and it still doesn't support nested properties well :( Not that it's impossible to do though...

> I am not really asking to fully support them, but it would be nice if a duplicate is returned when the property is pulled from the protoset.

Yeah, I guess it can be a solution for now. Way better than returning a reference to something in the protoset. I'll do some testing and put it into the next release if it doesn't interfere with anything.

**Comment by peter-kish** [comment]
Also adding the "bug" label to investigate those warnings from the first message.

**Comment by snipercup** [comment]
Thank you for looking into it. Working with ID's is perfect for my use case. If you want to see what I am working towards you can look at the json definition of another game, in this case some guns: https://github.com/CleverRaven/Cataclysm-DDA/blob/master/data/json/items/gun/9mm.json Those definitions do use type. I might end up using them after all, but I feel like my current idea gives more flexibility. One of the things I want to do is query the dataset for all "Magazine" items and add them to a list so you can select them when modifying a gun definition. 

After applying the workaround I didn't see those errors anymore.

**Comment by peter-kish** [comment]
> I am not really asking to fully support them, but it would be nice if a duplicate is returned when the property is pulled from the protoset.

Should be implemented in [v2.4.1](https://github.com/peter-kish/gloot/releases/tag/v2.4.1).
